src/Microsoft.ML.Data/Commands/CrossValidationCommand.cs

```diff
@@ -69,8 +70,8 @@ public sealed class Arguments : DataCommand.ArgumentsBase
             [Argument(ArgumentType.LastOccurenceWins, HelpText = "Whether we should cache input training data", ShortName = "cache")]
             public bool? CacheData;
 
-            [Argument(ArgumentType.Multiple, HelpText = "Transforms to apply prior to splitting the data into folds", ShortName = "prexf")]
-            public KeyValuePair<string, SubComponent<IDataTransform, SignatureDataTransform>>[] PreTransform;
+            [Argument(ArgumentType.Multiple, HelpText = "Transforms to apply prior to splitting the data into folds", ShortName = "prexf", SignatureType = typeof(SignatureDataTransform))]
+            public KeyValuePair<string, IComponentFactory<IDataView, IDataTransform>>[] PreTransform;
```

Comment:
This is not necessarily how I'd prefer to do these things, for two reasons. Firstly, while `IComponentFactory` is an interface, usually when appearing in these arguments objects we actually are indicating a specific inheriting interface appropriate for this task... e.g., I would expect to see something like `IDataTransformFactory`.

Also: our goal would be to also get rid of this little "signature" concept.

Signature is used for two major purposes, which I list here and describe how I think they could be better accomplished.

1. It constrains the type based on the generic type (in the case where signatures are used. e.g., we might imagine an `ITrainerFactory<IBinaryPredictor>` in the pre-estimator world). This would I think suffice.

2. It indicates what parameters will be necessary to pass in to instantiate the object in the `Create` method or the constructor. However, in the case where the signature is indicating "globally appropriate" input (in this case, `IDataTransform` is always instantiated with the input data), it would be better for the subinterface to just have a method *directly*, without using all this reflection based stuff at all.

I would have an actual subinterface type (that is, `IDataTransformFactory` or something similarly named. Then this interface would itself have a method to create the transform, and that method would have at least two parameters (in the current world): the `env` and the `input` data.

Reply:
> Also: our goal would be to also get rid of this little "signature" concept.

I 100% agree.  But it is too baked into the system right now to remove it in a single commit. Moving to `IComponentFactory` will lessen the dependence on the "signature" concept, but it will have to stay for a while until we can completely revamp the ComponentCatalog system.

> usually when appearing in these arguments objects we actually are indicating a specific inheriting interface appropriate for this task... e.g., I would expect to see something like IDataTransformFactory.

>I would have an actual subinterface type (that is, IDataTransformFactory or something similarly named. Then this interface would itself have a method to create the transform, and that method would have at least two parameters (in the current world): the env and the input data.

I'd need to think about this a bit more. My initial reaction is that this would cause the following disadvantages:

1. We would have a lot of different IComponentFactory interfaces instead of the 3 we have so far. As we added new types of components, we would constantly be needing to create new interfaces.
    - I think this is analogous with delegates in .NET 1.0 and 2.0.  If you needed a delegate that took in an `int` and returned a `bool`, you needed to make a brand new delegate type.  But in .NET 3.5, we added `Action<>` and `Func<>` delegates. And now you barely ever need to define custom delegate types.  Instead you can just use `Func<int, bool>`.
2. We couldn't have the small set of [`SimpleComponentFactory` classes](https://github.com/eerhardt/machinelearning/blob/05141a81a788d3fc863ba436791e9059865d081d/src/Microsoft.ML.Core/EntryPoints/ComponentFactory.cs#L32) that just takes a delegate to create the component.  We would have to define a new class for each subinterface.
3. Similarly as (2) above, the CmdParser class would have to (1) know about (2) define a new class, and (3) create different objects for each of the subinterfaces the API defines.
    - (Obviously I know the command line isn't a high priority, or selling reason, but we still need to support it and other tooling scenarios)

Reply:
/cc @Zruty0 - any thoughts here?

Reply:
I agree with @eric.erhardt@microsoft.com 's reasoning. I think we could check it in as is.

---
In reply to: [207046122](https://github.com/dotnet/machinelearning/pull/622#discussion_r207046122) [](ancestors = 207046122)

Reply:
Yea, I am OK with this solution. 

---
In reply to: [207057777](https://github.com/dotnet/machinelearning/pull/622#discussion_r207057777) [](ancestors = 207057777)